### ISSUE TITLE
BlockingMemoryStream Intermittent build failures

### ISSUE THREAD

**Issue by brthor** [dotnet/cli#2701]
I am seeing this intermittently in different places. 

Any ideas @eerhardt ?

```
[FATAL ERROR] System.InvalidOperationException
      System.InvalidOperationException : The collection has been marked as complete with regards to additions.
      Stack Trace:
           at System.Collections.Concurrent.BlockingCollection`1.TryAddWithNoTimeValidation(T item, Int32 millisecondsTimeout, CancellationToken cancellationToken)
           at System.Collections.Concurrent.BlockingCollection`1.Add(T item)
           at Microsoft.DotNet.Cli.Utils.BlockingMemoryStream.Write(Byte[] buffer, Int32 offset, Int32 count)
           at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
           at System.IO.SyncTextWriter.WriteLine(String value)
           at System.Console.WriteLine(String value)
```

**Comment by eerhardt** [comment]
Are you seeing this in CI or locally? Is it a test or manual usage?

The stream uses a BlockingCollection underneath the covers and calls CompleteAdding when it is done. However in this case it looks like someone is still trying to write to the stream.

We could add a check for IsAddingCompleted, but even in that case we should still throw since the stream is supposed to be in a state where no more writes are supposed to happen.

**Comment by brthor** [comment]
I am using `dotnet -v test` locally and seeing this intermittently, but also have seen it in random CI runs.

If it happens again, I'll try to get more details.

**Comment by brthor** [comment]
Repro Here:

http://dotnet-ci.cloudapp.net/job/dotnet_cli/job/rel_1.0.0/job/release_osx_x64_prtest/1676/

**Comment by eerhardt** [comment]
Perfect - that helped me figure out what is going on.

This is the same problem as #2768.  Basically the test doesn't expect any other test to write out to the console while it is running.  But in the link above, it shows:

```
16:44:35     Microsoft.DotNet.Cli.Utils.Tests.GivenAProjectDependenciesCommandFactory.It_resolves_tools_whose_package_name_is_different_than_dll_name [FAIL]
16:44:35       System.InvalidOperationException : The collection has been marked as complete with regards to additions.
16:44:35       Stack Trace:
16:44:35            at System.Collections.Concurrent.BlockingCollection`1.TryAddWithNoTimeValidation(T item, Int32 millisecondsTimeout, CancellationToken cancellationToken)
16:44:35            at Microsoft.DotNet.Cli.Utils.BlockingMemoryStream.Write(Byte[] buffer, Int32 offset, Int32 count)
16:44:35            at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
16:44:35            at System.IO.SyncTextWriter.WriteLine(String value)
16:44:35            at System.Console.WriteLine(String value)
16:44:35            at Microsoft.DotNet.TestFramework.TestInstance..ctor(String testAssetRoot, String testDestination)
16:44:35            at Microsoft.DotNet.TestFramework.TestAssetsManager.CreateTestInstance(String testProjectName, String callingMethod, String identifier)
16:44:35            at Microsoft.DotNet.Cli.Utils.Tests.GivenAProjectDependenciesCommandFactory.It_resolves_tools_whose_package_name_is_different_than_dll_name()
```

Which means this other test is writing to the console at the exact time the BuiltInCommandTest has swapped out the console text writer for its in memory blocking stream.

The same fix will fix this as well - mock out using the console in the BuiltInCommandTest.

**Comment by eerhardt** [comment]
This is fixed with #2769.
